src/libraries/System.Text.RegularExpressions/src/dotnet-regular-expressions.md

```diff
@@ -0,0 +1,237 @@
+* `RegexCompiler` uses the tables from this object, but generates its own code for the Boyer-Moore search
+
+### RegexCode
+
+* Abstract representation of the "program" for a particular pattern
+* Created by `RegexWriter`
+* Code is an array of integers. Within the array, op-codes' types are indicated by integer consts analogous to those on `RegexNode`.
+* Has several related data structures such as a string table, a captures table, and prefixes
+
+## Key types - Compilation (if not interpreted)
+
+### RegexCompiler (public abstract)
+
+* Responsible for compiling `RegexCode` to a `RegexRunnerFactory`
+* As implemented, uses `RegexLWCGCompiler`
+* Ha utility method `CompileToAssembly` that invokes `RegexParser` and `RegexWriter` directly then uses `RegexAssemblyCompiler` (see note for that type)
+* Key protected methods are `GenerateFindFirstChar()` and `GenerateGo()`
+* Created and used only from `RegexRunnerFactory Regex.Compile(RegexCode code, RegexOptions options...)`
+* Implements `RegexRunnerFactory RegexCompiler.Compile(RegexCode code, RegexOptions options...)`
+
+### RegexLWCGCompiler (is a RegexCompiler)
+
+* Creates a `CompiledRegexRunnerFactory` using `RegexRunnerFactory FactoryInstanceFromCode(RegexCode .. )`
+
+### RegexRunnerFactory (public pure abstract)
+
+* Reuseable: creates `RegexRunner`s on demand with `RegexRunner CreateInstance()`
+* Not relevant to interpreted mode
+* Must be thread-safe, as each `Regex` holds one, and `Regex` is thread-safe
+
+### CompiledRegexRunnerFactory (is a RegexRunnerFactory)
+
+* Created by `RegexLWCGCompiler`
+* Creates `CompiledRegexRunner` on request
+
+### RegexAssemblyCompiler
+
+* Created and used by `RegexCompiler.CompileToAssembly(...)` to write compiled regex to disk: at present, writing to disk is not implemented, because Reflection.Emit does not support it.
+
+## Key types - Execution
+
+### RegexRunner (public abstract)
+
+* Responsible for executing a regular expression: not thread-safe
+* Resueable: each call to `Scan(..)` begins a new execution
+* Lots of protected members: tracking position, execution stacks, and captures:
+  * `protected abstract void Go()`
+  * `protected abstract bool FindFirstChar()`
+  * `public Match? Scan(System.Text.RegularExpressions.Regex regex, string text...)` calls `FindFirstChar()` and `Go()`
+* Has a "quick" mode that does not instantiate any captures: used by `Regex.IsMatch(..)` which does not expose captures to the caller
+* Concrete instances created by `Match? Regex.Run(...)` calling either `RegexRunner CompiledRegexRunnerFactory.CreateInstance()` or newing up a `RegexInterpreter`
+
+### RegexInterpreter (is a RegexRunner)
+
+* See above. Note that this is sealed.
+
+### CompiledRegexRunner (is a RegexRunner)
+
+* See above.
+
+## Results
+
+### Match (public, is a Group)
+
+* Represents one match of the pattern: there may be several
+* Holds a `Regex` in order to call `NextMatch()`
+* Created by `RegexRunner`
```

Comment:
Probably worth calling out that while Regex is thread-safe, the things it produces (like Match) are not.